File: DSA/SearchingAlgorithms/Code/BinarySearch.cs
using System;
using System.Linq;

// Binary Search in C#
// Search for an element in a sorted array by dividing search space in half

class BinarySearch {
    
    static int BinarySearchMethod(int[] arr, int target) {
        int left = 0, right = arr.Length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (arr[mid] == target)
                return mid;
            else if (arr[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return -1;
    }
    
    static int BinarySearchRecursive(int[] arr, int left, int right, int target) {
        if (left > right)
            return -1;
        
        int mid = left + (right - left) / 2;
        
        if (arr[mid] == target)
            return mid;
        else if (arr[mid] < target)
            return BinarySearchRecursive(arr, mid + 1, right, target);
        else
            return BinarySearchRecursive(arr, left, mid - 1, target);
    }
    
    static int BinarySearchFirstOccurrence(int[] arr, int target) {
        int left = 0, right = arr.Length - 1;
        int result = -1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (arr[mid] == target) {
                result = mid;
                right = mid - 1;
            } else if (arr[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return result;
    }
    
    static int BinarySearchLastOccurrence(int[] arr, int target) {
        int left = 0, right = arr.Length - 1;
        int result = -1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (arr[mid] == target) {
                result = mid;
                left = mid + 1;
            } else if (arr[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return result;
    }
    
    static void DisplayArray(int[] arr) {
        Console.Write("Sorted Array: ");
        foreach (int num in arr)
            Console.Write(num + " ");
        Console.WriteLine();
    }
    
    static void Main() {
        // IMPORTANT: Array must be sorted for binary search
        int[] arr = {11, 12, 22, 25, 34, 64, 90};
        
        DisplayArray(arr);
        
        // Test 1: Search for element in middle
        int target = 25;
        Console.WriteLine("\nSearching for " + target + " (Iterative): ");
        int result = BinarySearchMethod(arr, target);
        if (result != -1)
            Console.WriteLine("Found at index " + result);
        else
            Console.WriteLine("Not found");
        
        // Test 2: Search using recursive approach
        Console.WriteLine("Searching for " + target + " (Recursive): ");
        result = BinarySearchRecursive(arr, 0, arr.Length - 1, target);
        if (result != -1)
            Console.WriteLine("Found at index " + result);
        else
            Console.WriteLine("Not found");
        
        // Test 3: Search for first element
        target = 11;
        Console.WriteLine("Searching for " + target + ": ");
        result = BinarySearchMethod(arr, target);
        if (result != -1)
            Console.WriteLine("Found at index " + result);
        else
            Console.WriteLine("Not found");
        
        // Test 4: Search for last element
        target = 90;
        Console.WriteLine("Searching for " + target + ": ");
        result = BinarySearchMethod(arr, target);
        if (result != -1)
            Console.WriteLine("Found at index " + result);
        else
            Console.WriteLine("Not found");
        
        // Test 5: Search for non-existing element
        target = 50;
        Console.WriteLine("Searching for " + target + ": ");
        result = BinarySearchMethod(arr, target);
        if (result != -1)
            Console.WriteLine("Found at index " + result);
        else
            Console.WriteLine("Not found");
        
        // Test 6: Using Array.BinarySearch()
        Console.WriteLine("\nUsing Array.BinarySearch():");
        target = 34;
        int index = Array.BinarySearch(arr, target);
        Console.WriteLine("Array.BinarySearch for " + target + ": " 
                         + (index >= 0 ? "Index " + index : "Not found"));
        
        // Test 7: First and last occurrences
        int[] arr2 = {11, 12, 12, 12, 22, 25, 34};
        Console.WriteLine("\nArray with duplicates: " + string.Join(" ", arr2));
        target = 12;
        Console.WriteLine("First occurrence of " + target + ": " 
                         + BinarySearchFirstOccurrence(arr2, target));
        Console.WriteLine("Last occurrence of " + target + ": " 
                         + BinarySearchLastOccurrence(arr2, target));
        
        // Test 8: Using LINQ
        Console.WriteLine("\nUsing LINQ:");
        target = 25;
        bool found = arr.Contains(target);
        Console.WriteLine("Contains " + target + ": " + found);
        var index2 = Array.FindIndex(arr, x => x == target);
        Console.WriteLine("FindIndex for " + target + ": " + (index2 >= 0 ? "Index " + index2 : "Not found"));
        
        // Complexity Analysis
        Console.WriteLine("\nComplexity Analysis:");
        Console.WriteLine("Time Complexity: O(log n) - Logarithmic");
        Console.WriteLine("Space Complexity: O(1) - Iterative, O(log n) - Recursive");
        Console.WriteLine("Best Case: O(1)");
        Console.WriteLine("Worst Case: O(log n)");
        Console.WriteLine("\nNote: Requires sorted array!");
    }
}

File: DSA/SearchingAlgorithms/Code/BinarySearch.c
#include <stdio.h>

// Binary Search in C
// Search for an element in a sorted array by dividing search space in half

int binarySearch(int arr[], int n, int target) {
    int left = 0, right = n - 1;
    
    while (left <= right) {
        int mid = left + (right - left) / 2;
        
        if (arr[mid] == target)
            return mid;  // Element found
        else if (arr[mid] < target)
            left = mid + 1;  // Search in right half
        else
            right = mid - 1;  // Search in left half
    }
    return -1;  // Element not found
}

// Recursive binary search
int binarySearchRecursive(int arr[], int left, int right, int target) {
    if (left > right)
        return -1;
    
    int mid = left + (right - left) / 2;
    
    if (arr[mid] == target)
        return mid;
    else if (arr[mid] < target)
        return binarySearchRecursive(arr, mid + 1, right, target);
    else
        return binarySearchRecursive(arr, left, mid - 1, target);
}

void displayArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    // IMPORTANT: Array must be sorted for binary search
    int arr[] = {11, 12, 22, 25, 34, 64, 90};
    int n = 7;
    int target;
    
    printf("Sorted Array: ");
    displayArray(arr, n);
    
    // Test 1: Search for element in middle
    target = 25;
    printf("\nSearching for %d (Iterative): ", target);
    int result = binarySearch(arr, n, target);
    if (result != -1)
        printf("Found at index %d\n", result);
    else
        printf("Not found\n");
    
    // Test 2: Search using recursive approach
    printf("Searching for %d (Recursive): ", target);
    result = binarySearchRecursive(arr, 0, n - 1, target);
    if (result != -1)
        printf("Found at index %d\n", result);
    else
        printf("Not found\n");
    
    // Test 3: Search for first element
    target = 11;
    printf("Searching for %d: ", target);
    result = binarySearch(arr, n, target);
    if (result != -1)
        printf("Found at index %d\n", result);
    else
        printf("Not found\n");
    
    // Test 4: Search for last element
    target = 90;
    printf("Searching for %d: ", target);
    result = binarySearch(arr, n, target);
    if (result != -1)
        printf("Found at index %d\n", result);
    else
        printf("Not found\n");
    
    // Test 5: Search for non-existing element
    target = 50;
    printf("Searching for %d: ", target);
    result = binarySearch(arr, n, target);
    if (result != -1)
        printf("Found at index %d\n", result);
    else
        printf("Not found\n");
    
    // Time Complexity Analysis
    printf("\nComplexity Analysis:\n");
    printf("Time Complexity: O(log n) - Logarithmic\n");
    printf("Space Complexity: O(1) - Iterative, O(log n) - Recursive\n");
    printf("Best Case: O(1) - Element at middle\n");
    printf("Worst Case: O(log n) - Element not found or at extreme\n");
    
    return 0;
}
